### ISSUE TITLE
Add possibility to output attentions from the ONNX models

### ISSUE THREAD

**Issue by jegork** [huggingface/optimum#325]
### Feature request

Hello!

I am interested in the possibility of retrieving not only the logits from the ONNX exported model, but also the attentions (which can be set via output_attentions=True in transformers). 
I have tried setting output_attentions=True in the config.json file of the model repository, but an exception is raised when the onnxruntime InferenceSession is initialized.

```python
Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from ort_model_optimized.onnx failed:/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-req-build-u762x_rp/onnxruntime/core/graph/graph.cc:1236 void onnxruntime::Graph::InitializeStateFromModelFileGraphProto() This is an invalid model. Graph output (attentions) does not exist in the graph
```

### Motivation

I am using attention scores for calculating the importance, but it is currently impossible to extract them in our production environment as we are using ONNX-exported transformers models.

### Your contribution

If you could guide me a bit on what to do, I could submit a PR.

**Comment by JingyaHuang** [comment]
Hi @jegork, 

As the default __OnnxConfig__ takes [common outputs](https://github.com/huggingface/transformers/blob/main/src/transformers/onnx/config.py#L76) which do not include `output_attentions`, the output of exported ONNX model will not contain `output_attentions`.

We are working on making the onnx configs more flexible, for the moment a workaround would be creating a wrapper of the onnx config to include `output_attentions` as a part of the output. Here is an example of a custom sbert onnx config:
```python
class CustomSTOnnxConfig(OnnxConfig, ABC):
    def __init__(self, config: OnnxConfig):
        self.__dict__ = copy.deepcopy(config.__dict__)
        self._onnx_config = config

    @property
    def inputs(self):
        return self._onnx_config.inputs

    @property
    def outputs(self):
        return OrderedDict({"last_hidden_state": {0: "batch", 1: "sequence"}, "pooler_output": {0: "batch"}})
```
Then you can wrap your onnx config with the wrapper
```python
custom_onnx_config = CustomSTOnnxConfig(onnx_config)
```

And after the export, if you want to leverage the ORTModels in optimum for inference, here is a [PR in progress](https://github.com/huggingface/optimum/pull/303). With the class `ORTModelForCustomTasks`, you will be able to put any custom onnx model for inference.

**Comment by jegork** [comment]
Thank you for the fast response @JingyaHuang

I have managed to export the model to ONNX, the only problem left is optimizing the model. When I export an optimized ONNX model via ORTOptimizer (optimization_level = 2), the attention nodes get disconnected from the graph (I guess the nodes that they were connected to get replaced).

Here on the visualization, you can see the output nodes disconnected:
<img width="1431" alt="image" src="https://user-images.githubusercontent.com/43540177/181766175-d5fa95b1-0a45-4512-87bf-acd6c49438ed.png">


Do you know if there is any way to keep the attentions together with the graph optimization from ORTOptimizer?

**Comment by jegork** [comment]
I have managed to isolate the problem. Apparently this happens after optimize_by_fusion is called. Thus, with optimize_with_onnxruntime_only=True for OptimizationConfig everything works correctly.

I provide an example for you to reproduce.

```python
from abc import ABC
from transformers.onnx.config import OnnxConfig
import copy
from collections import OrderedDict
from optimum.onnxruntime.configuration import OptimizationConfig
from optimum.onnxruntime import ORTOptimizer
from transformers import AutoModelForSequenceClassification, AutoTokenizer
import onnxruntime

class CustomOnnxConfig(OnnxConfig, ABC):
    def __init__(self, config: OnnxConfig):
        self.__dict__ = copy.deepcopy(config.__dict__)
        self._onnx_config = config

    @property
    def inputs(self):
        return self._onnx_config.inputs

    @property
    def outputs(self):
        return OrderedDict(
            **{"logits": {0: "batch"}},
            **{
                f"attentions_{i}": {
                    0: "batch",
                    1: "heads",
                    2: "sequence1",
                    3: "sequence2",
                }
                for i in range(self._onnx_config._config.num_hidden_layers)
            }
        )


model_name_or_path = "cross-encoder/mmarco-mMiniLMv2-L12-H384-v1"

model = AutoModelForSequenceClassification.from_pretrained(model_name_or_path, num_labels=1)
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)

model.config.output_attentions = True

optimizer = ORTOptimizer(tokenizer, model, "sequence-classification")

optimizer._onnx_config = CustomOnnxConfig(optimizer._onnx_config)
optimizer.model.config.save_pretrained("/tmp/model/")

optimizer.export(
    onnx_model_path="/tmp/model/ort_model.onnx",
    onnx_optimized_model_output_path="/tmp/model/ort_model_optimized.onnx",
    optimization_config=OptimizationConfig(optimization_level=2),
)

ort_session = onnxruntime.InferenceSession("/tmp/searcher/ort_model_optimized.onnx")
```

**Comment by JingyaHuang** [comment]
Hi @jegork , glad to hear that you have found a way to solve it. By setting `optimize_with_onnxruntime_only=True` you will turn off the whole [fusion optimization](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/optimizer.py#L252), have you tried precising [`FusionOptions`](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/fusion_options.py#L15) to locate the fusion leading to the failure? This might help enabling some other non-fatal fusion.

**Comment by jegork** [comment]
Hello @JingyaHuang, after taking a deeper look, it turned out that attention fusion was causing the aforementioned problem, so just using `OptimizationConfig(disable_attention=True)` worked well enough (although the attention fusion is not working, the remaining fusion optimization operations still work correctly).

Thank you for your time and guidance!
